File: plugins/git-status.30s.dart
#!/usr/bin/env dart
/// Git Status - Shows git repository status
import 'dart:io';

void main() async {
  // Check if in a git repo
  final gitCheck = await Process.run('git', ['rev-parse', '--is-inside-work-tree']);
  if (gitCheck.exitCode != 0) {
    print(' Not a repo');
    print('---');
    print('Not inside a git repository');
    return;
  }

  // Get current branch
  final branchResult = await Process.run('git', ['branch', '--show-current']);
  final branch = branchResult.stdout.toString().trim();

  // Get status
  final statusResult = await Process.run('git', ['status', '--porcelain']);
  final status = statusResult.stdout.toString().trim();
  final lines = status.isEmpty ? <String>[] : status.split('\n');

  final modified = lines.where((l) => l.startsWith(' M') || l.startsWith('M ')).length;
  final added = lines.where((l) => l.startsWith('A ') || l.startsWith('??')).length;
  final deleted = lines.where((l) => l.startsWith(' D') || l.startsWith('D ')).length;

  final changes = modified + added + deleted;
  final icon = changes > 0 ? '' : '';
  final color = changes > 0 ? 'orange' : 'green';

  print('$icon $branch | color=$color');
  print('---');
  print('Branch: $branch');
  print('---');

  if (changes > 0) {
    print('Changes:');
    if (modified > 0) print('  Modified: $modified');
    if (added > 0) print('  Added: $added');
    if (deleted > 0) print('  Deleted: $deleted');

    print('---');
    for (final line in lines.take(10)) {
      final file = line.substring(3);
      final type = line.substring(0, 2).trim();
      print('$type $file');
    }
    if (lines.length > 10) {
      print('...and ${lines.length - 10} more files');
    }
  } else {
    print('Working tree clean');
  }

  print('---');
  print('Refresh | refresh=true');
}

File: plugins/git-status/git-status.30s.lua
-- git-status.30s.lua
-- Git Status using Crossbar CLI

local function env(name, default)
    local value = crossbar.env(name, default)
    if value == nil or value == '' then
        return default
    end
    return value
end

local function env_bool(name, default)
    local value = env(name, default and 'true' or 'false')
    if value == nil then
        return default
    end
    value = tostring(value):lower()
    return value == 'true' or value == '1' or value == 'yes' or value == 'on'
end

local function env_num(name, default)
    local value = env(name, tostring(default))
    local num = tonumber(value)
    if num == nil then
        return default
    end
    return num
end

local function trim(value)
    if value == nil then return '' end
    return (value:gsub('^%s+', ''):gsub('%s+$', ''))
end

local is_mobile = crossbar.isMobile()
if is_mobile then
    print('git N/A | color=gray')
    print('---')
    print('Git status is desktop-only')
    print('---')
    print('Refresh | refresh=true')
    return
end

local max_files = env_num('GIT_MAX_FILES', 10)
if max_files < 1 then max_files = 1 end
if max_files > 200 then max_files = 200 end
local show_files = env_bool('GIT_SHOW_FILES', true)

local git_check = trim(crossbar.exec('git rev-parse --is-inside-work-tree 2>/dev/null'))
if git_check ~= 'true' then
    print('git Not a repo | color=gray')
    print('---')
    print('Not inside a git repository')
    print('---')
    print('Refresh | refresh=true')
    return
end

local branch = trim(crossbar.exec('git branch --show-current'))
if branch == '' then branch = 'detached' end

local status = crossbar.exec('git status --porcelain')
local changes = 0
local lines = {}

for line in (status or ''):gmatch('[^\n]+') do
    changes = changes + 1
    table.insert(lines, line)
end

local color = changes > 0 and 'orange' or 'green'

print('git ' .. branch .. ' | color=' .. color)
print('---')
print('Branch: ' .. branch)

if changes > 0 then
    print('Changes: ' .. tostring(changes))
    if show_files then
        print('---')
        local count = 0
        for _, line in ipairs(lines) do
            if count < max_files then
                print(line)
            end
            count = count + 1
        end
        if count > max_files then
            print('...and ' .. tostring(count - max_files) .. ' more files')
        end
    end
else
    print('Working tree clean')
end

print('---')
print('Refresh | refresh=true')
